File: services/otel-lib-ts/src/getters.ts
import { type Meter, type Tracer, metrics, trace } from "@opentelemetry/api";

/**
 * Convenience wrapper around `trace.getTracer`. Returns a no-op tracer when
 * the SDK has not been started (e.g. when OTEL_SDK_DISABLED=true), which is
 * the desired behavior; instrumentation code should never need to null-check.
 */
export function getTracer(name: string): Tracer {
  return trace.getTracer(name);
}

/**
 * Convenience wrapper around `metrics.getMeter`. Same no-op semantics as
 * `getTracer` when the SDK is not started.
 */
export function getMeter(name: string): Meter {
  return metrics.getMeter(name);
}

File: services/otel-lib/src/panakoes_otel/_getters.py
"""Convenience getters for tracers and meters.

Wrapping `opentelemetry.trace.get_tracer` / `opentelemetry.metrics.get_meter`
at the package level keeps consuming code provider-agnostic: services
import from `panakoes_otel` rather than reaching into the OTel API
directly. If we ever swap providers (e.g. add a custom decorator), the
change lands here.
"""

from __future__ import annotations

from opentelemetry import metrics, trace
from opentelemetry.metrics import Meter
from opentelemetry.trace import Tracer


def get_tracer(name: str) -> Tracer:
    """Return a `Tracer` for the named instrumentation scope.

    Calling before `configure()` is harmless: OTel returns a proxy
    tracer that no-ops until a real provider is installed.
    """
    return trace.get_tracer(name)


def get_meter(name: str) -> Meter:
    """Return a `Meter` for the named instrumentation scope.

    Same proxy semantics as `get_tracer`.
    """
    return metrics.get_meter(name)
